File: DAA/python_codes/n-queens.py
def print_board(board):
    for row in board:
        print(" ".join(row))
    print()

def solve_n_queens(n, first_queen_row, first_queen_col):
    col = set()
    pos_diag = set()
    neg_diag = set()
    solutions = []
    board = [["."] * n for _ in range(n)]

    def backtrack(row):
        if row == n:
            solutions.append(["".join(r) for r in board])
            return

        for c in range(n):
            if c in col or (row + c) in pos_diag or (row - c) in neg_diag:
                continue

            col.add(c)
            pos_diag.add(row + c)
            neg_diag.add(row - c)
            board[row][c] = "Q"

            backtrack(row + 1)

            col.remove(c)
            pos_diag.remove(row + c)
            neg_diag.remove(row - c)
            board[row][c] = "."

    board[first_queen_row][first_queen_col] = "Q"
    col.add(first_queen_col)
    pos_diag.add(first_queen_row + first_queen_col)
    neg_diag.add(first_queen_row - first_queen_col)

    backtrack(first_queen_row + 1)
    
    return solutions

if __name__ == "__main__":
    n = int(input("Enter the size of the board (n): "))
    first_queen_position = input("Enter the row and column for the first queen (0-indexed, separated by space): ")
    first_queen_row, first_queen_col = map(int, first_queen_position.split())

    if first_queen_row >= n or first_queen_col >= n:
        print("Invalid position for the first queen!")
    else:
        solutions = solve_n_queens(n, first_queen_row, first_queen_col)
        
        if not solutions:
            print("No solution exists with the first queen placed at the specified position.")
        else:
            print("One of the solutions is:")
            print_board(solutions[0]) 

File: DAA/n-queens.cpp
#include <bits/stdc++.h>
#include <vector>
using namespace std;

void print_board(const vector<vector<int>>& board) {
    int N = board.size();
    for (int row = 0; row < N; row++) {
        for (int col = 0; col < N; col++) {
            cout << (board[row][col] == 1 ? 'Q' : '.') << " ";
        }
        cout << endl;
    }
}

bool is_safe(const vector<vector<int>>& board, int row, int col, int init_row, int init_col) {
    int N = board.size();

    for (int i = 0; i < row; i++) {
        if (board[i][col] == 1) {
            return false;
        }
    }

    for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
        if (board[i][j] == 1) {
            return false;
        }
    }

    for (int i = row - 1, j = col + 1; i >= 0 && j < N; i--, j++) {
        if (board[i][j] == 1) {
            return false;
        }
    }

    if (col == init_col || abs(row - init_row) == abs(col - init_col)) {
        return false;
    }

    return true; 
}

bool solve(vector<vector<int>>& board, int row, int init_row, int init_col) {
    int N = board.size();

    if (row == N) {
        print_board(board);
        return true;
    }

    // Skip the row with the initial queen
    if (row == init_row) {
        return solve(board, row + 1, init_row, init_col);
    }

    for (int col = 0; col < N; col++) {
        if (is_safe(board, row, col, init_row, init_col)) {
            board[row][col] = 1; // Place the queen
            if (solve(board, row + 1, init_row, init_col)) {
                return true; // Solution found
            }
            board[row][col] = 0; // Backtrack
        }
    }
    return false; 
}

int main() {
    int N, init_row, init_col;
    cout << "Enter the size of the board (N): ";
    cin >> N;
    vector<vector<int>> board(N, vector<int>(N, 0));

    cout << "Enter the position for the first queen (row and column, 0-indexed): ";
    cin >> init_row >> init_col;

    if (init_row >= N || init_col >= N || init_row < 0 || init_col < 0) {
        cout << "Invalid position for the first queen!" << endl;
        return 1;
    }

    board[init_row][init_col] = 1;
    cout << "\nPlacing the first queen at position (" << init_row << ", " << init_col << ")" << endl;
    cout << "The initial board setup is:" << endl;
    print_board(board);
    cout << endl;

    if (!solve(board, 0, init_row, init_col)) {
        cout << "\nNo solution exists for the given starting position." << endl;
    }

    return 0;
}
